src/QMCWaveFunctions/Fermion/MultiSlaterDeterminantFast.cpp

```diff
@@ -827,32 +884,45 @@ void MultiSlaterDeterminantFast::evaluateDerivativesWF(ParticleSet& P,
       else
       //usingDETS
       {
-        ValueVector_t& detValues_up = Dets[0]->detValues;
-        ValueVector_t& detValues_dn = Dets[1]->detValues;
-        ValueType psiinv            = static_cast<ValueType>(PsiValueType(1.0) / psiCurrent);
+        ValueType psiinv = static_cast<ValueType>(PsiValueType(1.0) / psiCurrent);
         for (size_t i = 1; i < C->size(); i++)
         {
           int kk = myVars->where(i - 1);
           if (kk < 0)
             continue;
-          const size_t upC = (*C2node)[0][i];
-          const size_t dnC = (*C2node)[1][i];
-          ValueType cdet   = detValues_up[upC] * detValues_dn[dnC] * psiinv;
-          dlogpsi[kk]      = cdet;
+          ValueType cdet = psiinv;
+          for (size_t id = 0; id < Dets.size(); id++)
+          {
+            size_t spinC = (*C2node)[id][i];
+            // assume that evaluateLog has been called in opt routine before
+            ValueVector_t& detValues_spin = Dets[id]->detValues;
+            cdet *= detValues_spin[spinC];
+          }
+          dlogpsi[kk] = cdet;
         }
       }
     }
   }
 
-  // FIXME this needs to be fixed by SPF to separate evaluateDerivatives and evaluateDerivativesWF for otbital rotation matrix
-  Dets[0]->evaluateDerivativesWF(P, optvars, dlogpsi, *Dets[1], psiCurrent, *C, (*C2node)[0], (*C2node)[1]);
-  Dets[1]->evaluateDerivativesWF(P, optvars, dlogpsi, *Dets[0], psiCurrent, *C, (*C2node)[1], (*C2node)[0]);
+  if (Dets.size() != 2)
+  {
+    throw std::runtime_error(
+        "MultiSlaterDeterminantFast::evaluateDerivativesWF only compatible with two quantum particle types.");
+  }
+  else
+  {
+    // FIXME this needs to be fixed by SPF to separate evaluateDerivatives and evaluateDerivativesWF for orbital rotation matrix
+    Dets[0]->evaluateDerivativesWF(P, optvars, dlogpsi, *Dets[1], psiCurrent, *C, (*C2node)[0], (*C2node)[1]);
+    Dets[1]->evaluateDerivativesWF(P, optvars, dlogpsi, *Dets[0], psiCurrent, *C, (*C2node)[1], (*C2node)[0]);
```

Comment:
I think in the MultiDiractDeterminant, evaluateDerivativesWF should be split and move the two determinant interaction parts to MultiSlaterDeterminantFast. An MultiDiractDeterminant object should not be aware of the existence of other MultiDiractDeterminant objects

Reply:
Should this come in a future PR?

Reply:
Yes. Future.